### PR TITLE
MAINT: refactor ufunc iter operand flags handling

### PR DESCRIPTION
**Pull request by mattip**
In order for the inner loop of matmul to use BLAS, we need to be able to require contiguous input and output memory. Additionally, if there is memory overlap between the output and the input we must not write the results to the output array until looping is finished. `linalg` solves this problem by copying in the inner loop, which solves the first problem but not the second. `einsum` and `dot` solve this problem by not going through `PyUFunc_GenericFunction`, i.e. they are not true gufuncs. 

I can use a custom `ufunc->type_resolver` function for matmul to copy the inputs to contiguous memory and create an output array with writeback semantics, but there is nowhere to trigger resolution of those semantics after iterating. The most convenient thing would be to use iter operand flags to get the `NpyIter` to manage the writeback semantics for me. However, the `NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE` defeats that strategy, it tells `NpyIter` to avoid creating a writeback buffer under conditions that matmul requires a writeback.

I tried one approach to solving this in PR #11381, which simply removed the `NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE` flag from generalized ufuncs (ones that call `PyUFunc_GeneralizedFunction`). It seems people count on that behaviour, and so the PR was rejected.

This is an attempt at another approach, where the handling of the flags is pushed into the `ufunc->type_resolver` function, and I can override default behaviour in the matmul-specific one that anyhow I need in order to force input arrays to be contiguous.

In this first step, I create a generic function to set per-operand flags and use it in the three looping strategies. In later steps I will try to move the function into the generic `ufunc->type_resolver`.

### CODE REVIEW

**Comment by mhvk** [comment]
@mattip - I only had a glancing look so far, but I like this as a refactoring, even independently of the final goal.  Was the idea to do this over multiple PRs, or over multiple commits?  I think the former may be better (at least for this one), so ping me when this is no longer a WIP.

**Comment by mattip** [comment]
@mhvk I have gone about as far as I can go with this refactoring. I am not sure where in the code the calls to the new `_ufunc_setup_flags` function should be placed, I pushed the calls up to be as early as possible just to prove that the calls work and nothing else overrides the flags. 

Note that the function itself changed flag handling logic so that if `ufunc->op_flags[i]` is set, the call will not override any of the flags. Previously output `ufunc->op_flags` were never read, and input flags were `or`ed together. This is to allow the behaviour I need for matmul-as-a-ufunc (PR #11381 and the description above)

**Review by mhvk — COMMENTED** [review]
@mattip - The changes look good but I am a bit uncomfortable with the changes to the input flags (outputs are OK, since they were never used anyway). See longer note.

cc @eric-wieser as I think it would be good to have another look.

File: numpy/core/src/umath/ufunc_object.c [hunk 1]
```diff
@@ -302,6 +302,49 @@ _find_array_prepare(ufunc_full_args args,
     return;
 }
 
+NPY_NO_EXPORT void
+_ufunc_setup_flags(PyUFuncObject *ufunc, npy_intp op_in_flags,
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: fixing

File: numpy/core/src/umath/ufunc_object.c [hunk 2]
```diff
@@ -302,6 +302,49 @@ _find_array_prepare(ufunc_full_args args,
...
+                   npy_intp op_out_flags, npy_uint32 *op_flags)
+{
+    int nin = ufunc->nin;
+    int nout = ufunc->nout;
+    int nop = nin + nout, i;
+    /* Set up the flags */
+    for (i = 0; i < nin; ++i) {
+        op_flags[i] = ufunc->op_flags[i];
+        if (op_flags[i] == 0) {
+            op_flags[i] = op_in_flags |
```
[the remark this answers is not in the document]
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: As I stated in the justification for the pull request: there is curently no option for user-defined flags to override the flags here, and I definitely want to exclude NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE. I could also imagine a user who would want to exclude NPY_ITER_ALIGNED if their inner loop does not depend on element access being aligned. I have no strong opinion on whether the `op_in_flags` should be only the exceptions from the common set of flags or should be orred with a `NPY_UFUNC_DEFAULT_OP_FLAGS = NPY_ITER_READONLY | NPY_ITER_ALIGNED | NPY_ITER_OVERLAP_ASSUMED_ELEMENTWISE`

[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/core/src/umath/ufunc_object.c]: I guess the question is whether there are people out there relying on the flags they define being OR'd with the default ones; we ideally do not break their code.

**Comment by mattip** [comment]
Rebased off master to reduce codecov noise, squashed commits, removed possibly backward-incompatible behavior with input flags

**Comment by mattip** [comment]
@mhvk I left the input flag logic as it was previously

**Review by mhvk — APPROVED** [review]
With this, I think we're safe and can discuss the question of whether the logic for inputs should change separately. 

I'm approving, but will wait a day before merging to give @eric-wieser a chance to chime in.

**Comment by mattip** [comment]
@eric-wieser any thoughts?

File: numpy/core/src/umath/ufunc_object.c [hunk 3]
```diff
@@ -2457,6 +2435,15 @@ PyUFunc_GeneralizedFunction(PyUFuncObject *ufunc,
...
         goto fail;
     }
+    /*
+     * We don't write to all elements, and the iterator may make
+     * UPDATEIFCOPY temporary copies. The output arrays (unless they are
+     * allocated by the iterator itself) must be considered READWRITE by the
+     * iterator, so that the elements we don't write to are copied to the
+     * possible temporary array.
+     */
+    _ufunc_setup_flags(ufunc, NPY_ITER_COPY | NPY_UFUNC_DEFAULT_INPUT_FLAGS,
```
Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: lines < 80 characters.

File: numpy/core/src/umath/ufunc_object.c [hunk 4]
```diff
@@ -2840,16 +2799,15 @@ PyUFunc_GenericFunction(PyUFuncObject *ufunc,
...
-         * This checks whether a trivial loop is ok, making copies of
-         * scalar and one dimensional operands if that will help.
-         */
-        trivial_loop_ok = check_for_trivial_loop(ufunc, op, dtypes, buffersize);
-        if (trivial_loop_ok < 0) {
-            goto fail;
-        }
+    if (wheremask != NULL) {
+        /* Set up the flags. */
+        default_op_out_flags = NPY_ITER_NO_SUBTYPE | NPY_ITER_WRITEMASKED | NPY_UFUNC_DEFAULT_OUTPUT_FLAGS;
```
Comment (charris) [inline numpy/core/src/umath/ufunc_object.c]: Lines too long. also below.

Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: fixed

File: numpy/core/src/umath/ufunc_object.c (hunk 3 above)
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: fixed

File: numpy/core/src/umath/ufunc_object.c [hunk 5]
```diff
@@ -302,6 +302,47 @@ _find_array_prepare(ufunc_full_args args,
...
+    NPY_ITER_NO_BROADCAST | \
+    NPY_ITER_NO_SUBTYPE | \
+    NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE
+
+NPY_NO_EXPORT void
+_ufunc_setup_flags(PyUFuncObject *ufunc, npy_uint32 op_in_flags,
+                   npy_uint32 op_out_flags, npy_uint32 *op_flags)
+{
+    /*
+     * Note input flags are handled differently than ouput flags.
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: typo

File: numpy/core/src/umath/ufunc_object.c [hunk 6]
```diff
@@ -302,6 +302,47 @@ _find_array_prepare(ufunc_full_args args,
...
+
+#define NPY_UFUNC_DEFAULT_OUTPUT_FLAGS \
+    NPY_ITER_ALIGNED | \
+    NPY_ITER_ALLOCATE | \
+    NPY_ITER_NO_BROADCAST | \
+    NPY_ITER_NO_SUBTYPE | \
+    NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE
+
+NPY_NO_EXPORT void
+_ufunc_setup_flags(PyUFuncObject *ufunc, npy_uint32 op_in_flags,
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: This could maybe do with a docstring

File: numpy/core/src/umath/ufunc_object.c [hunk 7]
```diff
@@ -302,6 +302,47 @@ _find_array_prepare(ufunc_full_args args,
...
+        /*
+         * If READWRITE flag has been set for this operand,
+         * then clear default READONLY flag
+         */
+        if (op_flags[i] & (NPY_ITER_READWRITE | NPY_ITER_WRITEONLY)) {
+            op_flags[i] &= ~NPY_ITER_READONLY;
+        }
+    }
+    for (i = nin; i < nop; ++i) {
+        op_flags[i] = ufunc->op_flags[i] ? ufunc->op_flags[i] : op_out_flags;
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Are `op_flags` and `ufunc->op_flags` here ever the same pointer?

**Comment by eric-wieser** [comment]
> linalg solves this problem by copying in the inner loop, which solves the first problem but not the second

Does this second problem have a visible effect in `linalg`?

File: numpy/core/src/umath/ufunc_object.c [hunk 8]
```diff
@@ -1333,7 +1351,8 @@ execute_legacy_ufunc_loop(PyUFuncObject *ufunc,
                     NPY_ORDER order,
                     npy_intp buffersize,
                     PyObject **arr_prep,
-                    ufunc_full_args full_args)
+                    ufunc_full_args full_args,
+                    npy_uint32 *op_flags)
```
Comment (eric-wieser) [inline numpy/core/src/umath/ufunc_object.c]: Docstring for this function should be updated with this new argument

File: numpy/core/src/umath/ufunc_object.c (hunk 7 above)
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: No. `op_flags` is the per-call copy of `ufunc->op_flags`, updated for the actual call arguments and variation (`reduce`, `__call__`, `reduceat`)

**Comment by mattip** [comment]
> Does this second problem have a visible effect in `linalg`?

The second problem was overlap between input and output, which means we need to use writeback semantics or another technique to prevent writing to input. As far as I can tell, the problem does not exist in `linalg` functions in the first place, correct me if I am wrong.

File: numpy/core/src/umath/ufunc_object.c (hunk 5 above)
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: reworked entirely

File: numpy/core/src/umath/ufunc_object.c (hunk 8 above)
Comment (mattip) [reply numpy/core/src/umath/ufunc_object.c]: Reworked

**Comment by mattip** [comment]
ping

**Comment by mattip** [comment]
@eric-wieser can we merge this?

**Comment by mattip** [comment]
@eric-wieser can we merge this PR? It is blocking further progress with PR #11133 matmul-as-ufunc

**Comment by stefanv** [comment]
Gentle ping to see if we can get this one closed down.

**Comment by mattip** [comment]
Merging. @eric-wieser if you have further comments please raise them in a new issue

**Pull request merged**